### ISSUE TITLE
Implement Proxmox VE (`proxmoxve`) Support

### ISSUE THREAD

**Issue by b-** [coreos/fedora-coreos-tracker#1652]
# Implement Proxmox VE (`proxmoxve`) Support

This issue is to track support for the Proxmox VE hypervisor, which is a free and open source KVM+Qemu platform. 

The official reserved platform ID is `proxmoxve`, and the implementation is probably going to be mostly ripped from the `ibmcloud` provider to start with, as it uses the same standard for its own cloud-init data.

## During Development

Create PRs addressing the following:

- [x] [Ignition](https://github.com/coreos/ignition/) ([PR](https://github.com/coreos/ignition/pull/1910))
  - [x] Add userdata fetch
  - [ ] ~~If the platform supports it (unlikely), add userdata deletion~~
- [x] [Afterburn](https://github.com/coreos/afterburn/) ([PR](https://github.com/coreos/afterburn/pull/1023))
  - [x] (Cloud Only) Add relevant attributes
  - [x] (Cloud Only) Add SSH key support if available
  - [x] (Cloud Only) Add hostname support if available
  - [ ] ~~(Cloud Only) Add check-in if needed (unlikely)~~
- [x] [stream-metadata-go](https://github.com/coreos/stream-metadata-go) (https://github.com/coreos/stream-metadata-go/pull/89)
  - [ ] Add platform to the `Media` struct in `release/release.go`
  - [ ] Add supporting code for new platform to `toStreamArch` func in `release/translate.go`
  - [ ] (Cloud Only) Cloud Images need to have an `Images` field
- [ ] (Cloud Only) [stream-metadata-rust](https://github.com/coreos/stream-metadata-rust/) ([example PR](https://github.com/coreos/stream-metadata-rust/pull/16))
- [x] [fedora-coreos-tracker](https://github.com/coreos/fedora-coreos-tracker/) (https://github.com/coreos/fedora-coreos-tracker/pull/1971)
  - [ ] Update the metadata for the new platform
- [x] [coreos-assembler](https://github.com/coreos/coreos-assembler) (https://github.com/coreos/coreos-assembler/pull/4142)
  - [ ] Updated `cmd-generate-release-meta`
  - [ ] `cosa osbuild <platform>` works
- [ ] [fedora-coreos-config](https://github.com/coreos/fedora-coreos-config/)
  - [ ] Add a stanza to `platforms.yaml` if the system should use a serial console, or both serial and graphical consoles
- [x] [fedora-websites-3.0](https://gitlab.com/fedora/websites-apps/fedora-websites/fedora-websites-3.0/) (https://gitlab.com/fedora/websites-apps/fedora-websites/fedora-websites-3.0/-/merge_requests/1068)
  - [ ] Add friendly name for platform to `components/utilities/FpDownloadItem.vue`
  - [ ] Add artifact to `pages/coreos/download.vue`
  - [ ] Possibly add logo to `content/editions/coreos/home.yml`
- [x] [fedora-coreos-browser](https://github.com/coreos/fedora-coreos-browser) (https://github.com/coreos/fedora-coreos-browser/pull/52)
  - [ ] Add a list element for new platform in `browser/index.html`
- [x] [build pipeline](https://github.com/coreos/fedora-coreos-pipeline) (https://github.com/coreos/fedora-coreos-pipeline/pull/1170)
  - [ ] Add platform to the list found in `config.yaml` for building the new artifact
- [x] [fedora-coreos-docs](https://github.com/coreos/fedora-coreos-docs) ([example PR](https://github.com/coreos/fedora-coreos-docs/pull/377))
  - [ ] Add a `provisioning-<platform>.adoc` that walks through how to setup the new platform
  - [ ] Add an entry in the `modules/ROOT/nav.adoc` that points to new documentation

## At Release

1. Merge upstream changes and put out a release:
   - [x] Ignition
   - [x] Afterburn
1. Merge metadata changes:
   - [x] stream-metadata-go
   - [ ] stream-metadata-rust
   - [x] fedora-coreos-tracker
   - [x] fedora website
   - [x] fedora-coreos-browser
1. Release updated components
   - [x] Create and follow release checklist for [stream-metadata-go](https://github.com/coreos/stream-metadata-go/blob/main/docs/development.md#release-process) 
       - https://github.com/coreos/stream-metadata-go/issues/90
   - [x] Ensure that Dependabot has PRed stream-metadata-go into fedora-coreos-stream-generator and coreos-assembler.  Merge the update PRs.
     - This can be triggered manually by navigating to [fedora-coreos-stream-generator's Dependabot](https://github.com/coreos/fedora-coreos-stream-generator/network/updates/) and [coreos-assembler's Dependabot](https://github.com/coreos/coreos-assembler/network/updates) respectively, then clicking "Check for updates".
       - This might need to be done a few times, as Dependabot might not pick up tag changes for a few attempts after initial tagging.
     - https://github.com/coreos/coreos-assembler/pull/4152
   - [x] Create and follow release checklist for [fedora-coreos-stream-generator](https://github.com/coreos/fedora-coreos-stream-generator/blob/main/docs/development.md#release-process)
1. Merge the following changes:
   - [x] coreos-assembler
1. Wait for updates made to coreos-assembler to be propagated to latest container
   - [x] Download latest version of coreos-assembler container. Verify platform support functionality.
1. Merge changes for:
   - [x] Build pipeline
1. Wait for new images to reach stable then merge documentation.
   - [x] fedora-coreos-docs merged

**Comment by b-** [comment]
Prior conversation available at https://github.com/coreos/fedora-coreos-tracker/issues/736

**Comment by b-** [comment]
Afterburn implementation will have to be completely different from ibmcloud: they use a weird MIME-encoding format for SSH keys and such that seems pretty unique to them.

Since Ignition can provide ssh keys, maybe we can ignore Afterburn for this platform? If using the custom cloud-init support as I am suggesting then I don't think any other/extra metadata (such as the hostname) will available to the platform. 

If it means anything, my personal workflow for CoreOS on Proxmox so far is the following:

1. Deploy Fedora CoreOS VM using the `ibmcloud` image (or a `proxmoxve`-stamped image after https://github.com/coreos/ignition/pull/1790 is merged) and a very simple Ignition file that _doesn't_ actually create any users or set any SSH keys or passwords, but _does_ install the [cloud-init](https://src.fedoraproject.org/rpms/cloud-init) package
2. Power off the VM and remove the ignition disk
3. Set the VM as a template

Then, if I clone the VM template in Proxmox, it's ready for a Proxmox-style cloud-init provisioning data to install any necessary packages, create users, set SSH keys, and any other provisioning. This can actually be set via the Proxmox GUI.

A negative side effect of doing the above and not providing any authentication information via Ignition or Afterburn (and then essentially replacing Afterburn with cloud-init) is that the VM _will_ warn you that Ignition and Afterburn weren't provided with any provisioning data.

Another negative side effect is that, of course, Ignition is the RedHat/Fedora/CoreOS/Flatcar way, not cloud-init. But it works -- I've deployed CoreOS (as well as [ublue-os/ucore](https://github.com/ublue-os/ucore)) this way for simple homelab workloads, and they're solid as ever

**Comment by jlebon** [comment]
Re. Afterburn support, there's a pending PR for it already in https://github.com/coreos/afterburn/pull/1023, so we can just link to that from the checklist. It hasn't been reviewed in depth yet, but I don't see anything about having to decode SSH keys there.

Re. the cloud-init UX, we definitely want to avoid users jumping through those hoops. Ignition and Afterburn should be able to read from the same metadata sources that cloud-init does to provision the node (i.e. files in attached config drive). From my understanding of https://github.com/coreos/fedora-coreos-tracker/issues/736, the Proxmox UI doesn't make it easy to pass !cloud-init userdata, but it is possible on the command line (see https://github.com/coreos/fedora-coreos-tracker/issues/736#issuecomment-776073906).

**Comment by b-** [comment]
Indeed, Proxmox only allows you to pass arbitrary data using the command line. 

https://github.com/coreos/afterburn/pull/1023 is reading Proxmox-style cloud-init data anyway, rather than Ignition data shoehorned into place as I was suggesting. I think that this would make a much better user experience, at the expense of being more limited than arbitrary Ignition data.

The only reason I didn't suggest doing this was because I figured that interpreting the cloud-init data sources might be out-of-scope for the Ignition project, and of course it's more limiting, but I do think it would provide a better UX.

Perhaps both could be supported? In other words, [if the first line of `user-data` is `#cloud-config`, interpret the data source as `cloud-init` yaml](https://github.com/enix/afterburn/blob/proxmoxve/tests/fixtures/proxmoxve/static/user-data). Otherwise, try to interpret it as Ignition json data.

**Comment by jdoss** [comment]
I am looking to start using FCOS on a test Proxmox VM host. Is there anything I can do to help out with this endeavor? It looks like  https://github.com/coreos/ignition/pull/1790 needs testing and review. @dustymabe if you or @jlebon can point me in the right direction for doing the testing, I offer myself as tribute. I should have my test Proxmox server online in a few weeks.

**Comment by jlebon** [comment]
@b- I presume https://github.com/coreos/fedora-coreos-tracker/issues/736#issuecomment-1780146794 is still the latest state of things? I.e. that there is no sane way for users to provide arbitrary user-data to the instance without jumping through various hoops?

So obviously, ideally we would enhance Proxmox to allow that more easily. 

I'm confused though, putting aside the CLI/API workarounds, if Proxmox fully owns the generated cloud-init, how do users provide their own cloud-init to do whatever else they want? Or is there a section of the web UI where they can type in extra cloud-init configs that get merged in? (Or is the expectation that you basically always just follow up with some config mgmt tool?)

**Comment by jlebon** [comment]
So basically if I'm understanding this right:
- it's very difficult currently to set the Ignition config as the user-data
- by default, Proxmox will set a cloud-init user-data and use that as its mechanism for SSH keys and hostname configuration

So Proxmox essentially uses cloud-init user-data as its platform metadata mechanism. Which... isn't great but we can live with I guess.

> Perhaps both could be supported? In other words, [if the first line of user-data is #cloud-config, interpret the data source as cloud-init yaml](https://github.com/enix/afterburn/blob/proxmoxve/tests/fixtures/proxmoxve/static/user-data?rgh-link-date=2024-01-24T22%3A07%3A48Z). Otherwise, try to interpret it as Ignition json data.

I'm kinda thinking actually that we just keep this in Afterburn and don't add any Ignition support for this at all for now if we don't really expect users to jump through the hoops necessary to set an Ignition config. (Though if there's an RFE somewhere in progress to have Proxmox make it easier to pass your own user-data, that changes things.)

Afterburn would just parse the strict subset of the cloud-init YAML schema that Proxmox uses. If it's somehow possible for users to pass in some custom cloud-init snippets that get merged in, then we should error out so we don't pretend to understand all of cloud-init's schema.

**Comment by dustymabe** [comment]
maybe we could hook into the `cicustom` step? 

https://pve.proxmox.com/wiki/Cloud-Init_Support#_custom_cloud_init_configuration

**Comment by b-** [comment]
> So basically if I'm understanding this right:
> 
> - it's very difficult currently to set the Ignition config as the user-data
> 
> - by default, Proxmox will set a cloud-init user-data and use that as its mechanism for SSH keys and hostname configuration
> 
>
>
> So Proxmox essentially uses cloud-init user-data as its platform metadata mechanism. Which... isn't great but we can live with I guess.

That is correct.

> I'm kinda thinking actually that we just keep this in Afterburn and don't add any Ignition support for this at all for now if we don't really expect users to jump through the hoops necessary to set an Ignition config. (Though if there's an RFE somewhere in progress to have Proxmox make it easier to pass your own user-data, that changes things.)
> 
> 
> 
> Afterburn would just parse the strict subset of the cloud-init YAML schema that Proxmox uses. If it's somehow possible for users to pass in some custom cloud-init snippets that get merged in, then we should error out so we don't pretend to understand all of cloud-init's schema.

I like this idea a lot. I think this probably would be the simplest path forward to making Proxmox-flavored CoreOS images that anybody could use.

**Comment by dustymabe** [comment]
also of interest - some discussion of support for fw_cfg upstream: https://bugzilla.proxmox.com/show_bug.cgi?id=4068#c5

**Comment by b-** [comment]
> maybe we could hook into the `cicustom` step? 
> 
> 
> 
> https://pve.proxmox.com/wiki/Cloud-Init_Support#_custom_cloud_init_configuration

This is what I was suggesting as an additional/alternative to parsing the Proxmox-flavored subset of the cloud-init yaml. 

The nice things about this approach is that it could also work on bare metal, as well as any other hypervisor, if one just follows the directions. 

Perhaps maybe a separate "baremetal" target could be made that would accept an ignition file passed via cicustom or on a USB stick/CD drive with the proper volume name?

**Comment by b-** [comment]
> @b- I presume https://github.com/coreos/fedora-coreos-tracker/issues/736#issuecomment-1780146794 is still the latest state of things? I.e. that there is no sane way for users to provide arbitrary user-data to the instance without jumping through various hoops?

That is correct

> I'm confused though, putting aside the CLI/API workarounds, if Proxmox fully owns the generated cloud-init, how do users provide their own cloud-init to do whatever else they want? Or is there a section of the web UI where they can type in extra cloud-init configs that get merged in? (Or is the expectation that you basically always just follow up with some config mgmt tool?)


Custom cloud-init metadata can be provided via one of two ways: 

The first way, which is the `cicustom` way, is that a "snippet"^[1] can be created on/uploaded to the server's storage (or a mountpoint on it). Then a CLI/API command (`-cicustom` on the CLI) can be used to generate the cloudinit drive and mount it on the VM.

The second way is to manually create a cloud-init drive. 

To be clear (and I'm sorry if I keep repeating myself), the way Proxmox actually supplies the cloud-init metadata is by creating an ISO9660 image with a given volume name and a few files with specific filenames on it.

I only mention this to make it clear why I think that a separate target format ("baremetal") supporting a custom ignition config provided that way as well would be nice: 

You could manually (or with a simple script) build an image or physical disk following that format, and then provide that to either any generic VM or baremetal PC, without needing to also supply boot args.

[1]: Proxmox terminology for text file like an ignition file, or a shell script ("hook script") to be run on VM operations

**Comment by jlebon** [comment]
OK, so it sounds like the `--cicustom` approach is quite advanced/off the beaten path and requires additional Proxmox privileges.

How does this compare to the "attach your own ISO" approach? Does that require any special privileges? Is it part of the regular UX of a Proxmox user? (I.e. how common is it to have to resort to generating your own ISO to attach to your VM?)

I think it'd be nice to support passing in Ignition configs, but it sounds to me like we need to optimize for the Afterburn cloud-init case instead to better match expectactions (i.e. that'd be what we primarily document, and the custom Ignition case would be an advanced section).

**Comment by abuisine** [comment]
`--cicustom` is the option flag for the `qm` binary accessible from a shell on a proxmoxve hypervisor, but you got a similar option through the JSON/HTTP API.

The main problems related to building a custom iso (for instance through terraform on a station) are the **upload** to the proxmoxve cluster as well as the **storage medium** configured at the cluster level to receive such file.

Should you want to stick with custom iso, it would require:
- configure a replicated storage backend (the iso is attached to the vm, and the vm can move from one hypervisor to another, so the iso must be accessible from any hypervisor)
- upload the iso, the preferred method is usually ssh/scp to one hypervisor on the replicated storage

Or you can leverage from proxmoxve **snippets**:
- snippets are, by default, replicated within a proxmoxve cluster
- upload of custom configuration file as a snippet can be done either through the HTTP based API or through ssh/scp or through the GUI
- through the cicustom configuration, proxmoxve handle by himself the iso generation 

I am very much in favor of the later option for two reasons:
1. ease of use (no preliminary complex configuration)
2. it seems also more inline with the proxmoxve way of dealing with specific configurations (snippets)

but I admit that it requires a modification of ignition in order to look at the right location (as proposed in https://github.com/coreos/ignition/pull/1844)

**Comment by b-** [comment]
@abuisine are you certain that snippets can be uploaded via the API? That's the snag I got caught on when working on this a few months ago — I could not find an API call to actually upload or create a snippet. I looked pretty thoroughly… 

If there _is_ such an API call (and by all means there _should_ be one), I absolutely agree that snippets + cicustom are the right way for this.

**Comment by abuisine** [comment]
Hi, let me double check all this and get back to you.

**Comment by b-** [comment]
For the reasons mentioned above, I do think that supporting the Proxmox subset of cloud-init would still be more useful to more people than ignition-as-cicustom data. Ideally both would be supported (possibly in their own separate flavors), but _It Would Be Nice_™ to be able to spin up a CoreOS VM on Proxmox without messing around with anything, too…

**Comment by abuisine** [comment]
Can you confirm the following:

Support of _the Promox subset of cloud-init_ only requires an afterburn modification.
An example is available here https://github.com/coreos/afterburn/pull/1023

Support of _ignition-as-cicustom data_ requires an ignition modification and is probably less useful.
Two PRs today https://github.com/coreos/ignition/pull/1790/files and https://github.com/coreos/ignition/pull/1844.
I propose to dismiss the one we created and to contribute to yours ? Does it make sense ?

For the _CoreOS VM on Proxmox without messing around with anything_ ... apart from DHCP I guess we are far from the goal :P

**Comment by jlebon** [comment]
Can someone provide an example of a Proxmox-provided cloud-init config? Is e.g. https://github.com/enix/afterburn/blob/proxmoxve/tests/fixtures/proxmoxve/dhcp/user-data (from https://github.com/coreos/afterburn/pull/1023) an accurate example? (Weird that they would provide both `user:` and `users:`; the latter I think overriding the former.)

The thing is there are keys in there that Afterburn is not really designed to handle and I don't think we should add support for. Basically of the user-data we should just support `hostname`/`fqin` and `ssh_authorized_keys` and that's it. And then document this clearly.

Re. `network-config`, unfortunately Afterburn's network configuration story hasn't had a lot of effort put into it due to the fact that it outputs systemd-networkd units and FCOS doesn't ship that. It'd be great to add support for outputting nmstate config files though.

Anyway, short-term the problem there is that we'll simply not be able to apply the network config on FCOS, which means that... yeah just DHCP will work. We should document that limitation too.

I think for completeness we indeed should also add support in Ignition for advanced users. In that case, (1) Ignition should no-op if the user-data starts with `#cloud-config` (otherwise assume it's Ignition and error out as usual if it's not), and (2) Afterburn should no-op if the user-data doesn't start with `#cloud-config`.

**Comment by arcln** [comment]
The config you'll find in Afterburn tests (the one you linked) was directly copied from Proxmox VE and is accurate.

For the `network-config`, it's true that even after implementing support in Afterburn, it wasn't working and we had to add our own systemd unit to invoke `afterburn --network-units`.

Finally, our Afterburn PR looks for `#cloud-config` header and will ignore the file if it doesn't start with that. We only miss this behaviour in Ignition, and we should be good to go.

**Comment by b-** [comment]
>(Weird that they would provide both user: and users:; the latter I think overriding the former.)

Actually the other way around, at least on an Ubuntu 22.04 VM I provisioned on my home server. I set a username in the Proxmox Web UI, and such a user was created that belongs to the `sudo` group with the attached SSH key.

If you leave the `User` field in the web UI empty, it shows `Default` in its place and the only difference in the actual cloud-init user-data is `/user:/d`, i.e., there's no `user: dummy-user` line in the `user-data` file.

**Comment by jlebon** [comment]
OK gotcha. In that case, we can't support that either. I think we can basically say that we'll always create the `core` user regardless of the set `User` in the web UI.

Ideally, we could add metadata to the image somehow that Proxmox could query to know what is/isn't supported and reflect that in the web UI. But ideally ideally, Proxmox would have first-class support for Ignition and/or custom user-data. :)

**Comment by b-** [comment]
> I think we can basically say that we'll always create the `core` user regardless of the set `User` in the web UI.

I think this is reasonable, as long as setting the `User` field to `core` (or empty) will cause attached SSH keys and passwd hashes to be applied to the `core` user.

> Ideally, we could add metadata to the image somehow that Proxmox could query to know what is/isn't supported and reflect that in the web UI. 

I imagine this might be possible in "the future™," at least because Proxmox's LXC support relies on a similar metadata to pick some distro-specific behavior. 

> But ideally ideally, Proxmox would have first-class support for Ignition and/or custom user-data. :)

Hard agree. Honestly, I wish they could at least do what Oracle Cloud's web UI does, which is a big text box whose contents becomes the `user-data` file. Oracle's Ignition and afterburn support are so simple, but also so close to (yet so far from) what Proxmox does… 😕

@jlebon would it be useful for me to attach a `network-config` with static settings?

**Comment by jlebon** [comment]
> > I think we can basically say that we'll always create the `core` user regardless of the set `User` in the web UI.
> 
> I think this is reasonable, as long as setting the `User` field to `core` (or empty) will cause attached SSH keys and passwd hashes to be applied to the `core` user.

Yeah, ignoring it will give that behaviour. (We always create the `core` user by default and you'd need to be able to specify an Ignition config in order to override that.)

> @jlebon would it be useful for me to attach a `network-config` with static settings?

I think there's an example config in the Afterburn PR tests: https://github.com/enix/afterburn/blob/proxmoxve/tests/fixtures/proxmoxve/static/network-config

**Comment by jlebon** [comment]
Ignition support was merged in https://github.com/coreos/ignition/pull/1910. So now we need a new Ignition release with that in FCOS, but it's not too early for someone to start working on the docs part.

**Comment by jdoss** [comment]
@jlebon @dustymabe is there anything that I can do to help get this pushed over the line? I would love to run FCOS on Proxmox myself and I am down to help but I am not exactly sure where to start.

**Comment by b-** [comment]
I wanted to build or customize a CoreOS image to use this new provider, but I couldn't figure out how to in the first place… 

has anyone successfully done so yet?

**Comment by b-** [comment]
> > is there anything that I can do to help get this pushed over the line? I would love to run FCOS on Proxmox myself and I am down to help but I am not exactly sure where to start.
> 
> The checklist is in the top comment. The next step is to write docs. The tricky part is that we don't have good tooling yet to restamp the platform ID, but it's certainly possible today. See e.g. [the proposed docs for Oracle Cloud](https://github.com/coreos/fedora-coreos-docs/pull/652), which have a gnarly script.

Thanks for linking that @jlebon. I'll check that out and try to test some things locally. 

Here's a [direct link to the Oracle Cloud docs file](https://github.com/travier/fedora-coreos-docs/blob/main-prov-oraclecloud/modules/ROOT/pages/provisioning-oraclecloud.adoc), which has those conversion scripts. They are indeed pretty gnarly, what with guestfish and the like, but I think I could make a GitHub Actions pipeline that runs through them and uploads the binaries, to sort of throw together a "build server" to restamp images and start actually testing this thing… shouldn't be too complicated

**Comment by jlebon** [comment]
@b- Nice! Thanks for detailing that. If you're feeling generous, consider opening a docs PR with that info. :)

Re. coreos-restamper, it could be interesting to clean it up a bit, containerize it, move it to the coreos org (if you're interested), and then reference that in our docs as the tooling for emerging platforms. WDYT (and others)?

**Comment by b-** [comment]
> Re. coreos-restamper, it could be interesting to clean it up a bit, containerize it, move it to the coreos org (if you're interested), and then reference that in our docs as the tooling for emerging platforms. WDYT (and others)?

Sure! That sounds great! I'd love to!

**Comment by jbtrystram** [comment]
This was discussed during today FCOS community meeting. 

We agreed to make proxmox a toplevel supported platform for Fedora CoreOS.
It was also discussed how we could build on @b- work to make a ready to use restamper container, document and maintain it. 
However we ran out of time and will continue the discussion next week. @b- and any other interested in this topic feel free to join  !
Meetings happen on Wednesdays 17:30 UTC, in matrix.  Join our [public channel](https://matrix.to/#/#coreos:fedoraproject.org) where we announce meetings details

**Comment by jdoss** [comment]
@b- @jlebon I'd be down to clean up and containerize Bri's work but I don't want to step on any toes here. Let me know if I can help and I can get a PR going.

**Comment by b-** [comment]
> 
> This was discussed during today FCOS community meeting. 
> 
> We agreed to make proxmox a toplevel supported platform for Fedora CoreOS.
> It was also discussed how we could build on @b- work to make a ready to use restamper container, document and maintain it. 
> However we ran out of time and will continue the discussion next week. @b- and any other interested in this topic feel free to join  !
> Meetings happen on Wednesdays 17:30 UTC, in matrix.  Join our [public channel](https://matrix.to/#/#coreos:fedoraproject.org) where we announce meetings details  

Thanks for the invite, I'd love to participate! I'm going to make an event in my calendar.

@jdoss I started a very rough containerization of that at [this repo](https://github.com/perchnet/coreos-restamper-container)

---

By the way, I realized when I tried to test my restamped image that the proxmoxve provider still needs quite a lot of work, in that Ignition and Afterburn are going in completely different directions as far as what they expect.

To be frank, Ignition is looking for a completely different kind of config than Afterburn is looking for, lol. 

Ignition's proxmoxve provider mounts a cloud-init disk the way Proxmox generates one, reads the `user-data` file, and configures the VM if it's actually ignition data in there. 

On the other hand, Afterburn's proxmoxve provider also mounts a cloud-init disc, but then it takes values from a "_real_" cloud-init user-data file, and complains loudly if it doesn't get all of the data it expects. 

I think this is because there doesn't really seem to be a very clear direction or consensus on how such data should be provided from Proxmox VE to a CoreOS VM. Which I don't blame, because any way I can imagine solving this problem is kind of a little funky/hacky.

I'm personally leaning slightly towards the approach of shoving an ordinary ignition config into a cloud-init disc as the Ignition provider currently expects, but we'd need to figure out the best happy path to encourage the Proxmox VE hypervisor to supply ignition data (i.e., its `cicustom` option), and what to supply to Afterburn if necessary for subsequent reboots. (This might not be as important)

This all became apparent to me when I tried manually creating and uploading an ISO with the ignition data as a `user-data` file using Linux CLI tools (rather than using Proxmox's `cicustom` setting to replace the `user-data` file in a Proxmox-generated cloud-init image with an ignition file): Since my custom image didn't supply `instance-id` (as Proxmox usually does with cloud-init) Afterburn complained loudly during boot.

There also, of course, is the possibility of just doing something completely out of left field for Proxmox, like using an rpm-ostree layer (via a containerfile with e.g., a CoreOS `FROM` directive) to just install Fedora's own RPM package of `cloud-init`, as Ubuntu does out-of-the-box. (i.e., `RUN rpm-ostree install cloud-init` Then you can ignore ignition and Afterburn entirely, and just provision it with cloud-init data. 

I've tested this before (on a Qeum image without restamping), and the only issue I found with _that_ is a message on the login prompt warning me that Ignition didn't create any users, which doesn't matter if cloud-init created them instead, lol.

**Comment by jdoss** [comment]
I was able to distilled the wonderful instructions that @b- created to a handful of shell commands. This does not include creating the  proxmoxve qcow2. I like using templates with my Proxmox server so the first block creates a template:

```bash
pvesm add dir snippets  --path /var/snippets --content snippets
qm create 101 --name fedora-coreos-stable-template --memory 2048 --net0 virtio,bridge=vmbr0,tag=1000 --scsihw virtio-scsi-pci
qm set 101 --scsi0 'local-lvm:0,import-from=/var/fcos/work/fedora-coreos-41.20250117.3.0-proxmoxve.x86_64.qcow2'
qm set 101 --ide2 local-lvm:cloudinit
qm set 101 --boot order=scsi0
qm set 101 --serial0 socket --vga serial0
qm template 101
```

and then we can launch a clone with this:

```bash
qm clone 101 1001 --name fcos-stable-test
qm resize 1001 scsi0 +90G
qm set 1001 --sshkey ${HOME}/.ssh/authorized_keys
qm set 1001 --ipconfig0 ip=192.168.1.10/24,gw=192.168.1.1
qm set 1001 --cicustom user=snippets:snippets/testing.json
qm set 1001 --ciupgrade 0
```

You can see everything with this:

```bash
qm cloudinit dump 1001 user
```

There are some issues however. When setting a static IP address the network doesn't get configured. 

```bash
[root@localhost etc]# ip a s
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute
       valid_lft forever preferred_lft forever
2: ens18: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether bc:24:11:5f:6c:1b brd ff:ff:ff:ff:ff:ff
    altname enp0s18
```

Looking at afterburn.service, it is apart of `metadata.target` and this doesn't exist on the qemu or openstack images (I tried a different image to see if I could get different results). Does `afterburn.service` need to run to configure the network with a static IP address? 

It can access the cloud-config drive just fine: 

```[root@localhost etc]# afterburn --provider proxmoxve --network-units /tmp/
Feb 06 05:11:43.645 WARN user-data does not have the expected header `#cloud-config`, ignoring this file
[root@localhost etc]# cat /tmp/20-eth0.network
[Match]
Name=eth0
MACAddress=bc:24:11:5f:6c:1b

[Network]
DNS=8.8.8.8

[Address]
Address=192.168.1.10/26

[Route]
Destination=0.0.0.0/0
Gateway=192.168.1.1
```

**Comment by jdoss** [comment]
It looks like https://github.com/coreos/afterburn/pull/1023 implemented systemd networkd support for Flatcar but not support for NetworkManager so this won't ever work as far as I can tell for configuring the network in FCOS. 

I tried to see if I could pass some kargs via 

```
qm set ${VM_ID} --args "-append 'ip=192.168.1.10::192.168.1.1:255.255.255.192:testing.example.com:ens18:none:8.8.8.8'"
```

but you cannot use `-append`  without setting `-kernel` which is not ideal. I think support for writing out NetworkManager config files via Afterburn will need to be added before Proxmox can be a first class platform for FCOS.

**Comment by b-** [comment]
> It looks like https://github.com/coreos/afterburn/pull/1023 implemented systemd networkd support for Flatcar but not support for NetworkManager so this won't ever work as far as I can tell for configuring the network in FCOS. 
> 
> I tried to see if I could pass some kargs via 
> 
> ```
> qm set ${VM_ID} --args "-append 'ip=192.168.1.10::192.168.1.1:255.255.255.192:testing.example.com:ens18:none:8.8.8.8'"
> ```
> 
> but you cannot use `-append`  without setting `-kernel` which is not ideal. I think support for writing out NetworkManager config files via Afterburn will need to be added before Proxmox can be a first class platform for FCOS.

That sounds reasonable. I need to look into how NetworkManager does config files (I'm guilty of using `nmtui` whenever I needed to change something with NetworkManager…), but I imagine it shouldn't be too difficult. 

Maybe Afterburn can/should have an abstraction of "backends" for certain features, to allow for different implementations like this. 

As an aside, Proxmox generally doesn't encourage the `args` parameter, because amongst other things it isn't checked for correctness and certain QEMU/KVM arguments can lead to security issues or otherwise completely change behavior of things. 

As a result, you can't normally add, remove, change, or even view an args parameter unless you have root access to the server, and still it only works on either the CLI or API (or manually editing the config files).

Also, the reason the qemu `-append` arg requires `-kernel` is because the only way it can append arguments to the kernel is by skipping the boot loader and booting the kernel directly, which Proxmox also doesn't "support," instead opting to always boot VMs using emulated BIOS (SeaBIOS) or UEFI (Tianocore/OVMF), both of which sort of imply a more traditional boot loader like grub or systemd-boot (or a UKI)

**Comment by jdoss** [comment]
> That sounds reasonable. I need to look into how NetworkManager does config files (I'm guilty of using `nmtui` whenever I needed to change something with NetworkManager…), but I imagine it shouldn't be too difficult.

I think the fast path would be to add support for writing out files to `/etc/NetworkManager/system-connections/` via Afterburn.

> As an aside, Proxmox generally doesn't encourage the `args` parameter, because amongst other things it isn't checked for correctness and certain QEMU/KVM arguments can lead to security issues or otherwise completely change behavior of things.
> 
> As a result, you can't normally add, remove, change, or even view an args parameter unless you have root access to the server, and still it only works on either the CLI or API (or manually editing the config files).
> 
> Also, the reason the qemu `-append` arg requires `-kernel` is because the only way it can append arguments to the kernel is by skipping the boot loader and booting the kernel directly, which Proxmox also doesn't "support," instead opting to always boot VMs using emulated BIOS (SeaBIOS) or UEFI (Tianocore/OVMF), both of which sort of imply a more traditional boot loader like grub or systemd-boot (or a UKI)

Yeah understand this was a bit hacky. I was just trying to work around this issue so I didn't have to inject the NetworkManger configs into user supplied ignition configs.

**Comment by jdoss** [comment]
Another thing I noticed in the cidata drive there is a vendor-data file that is empty and you can run 

```
qm set ${VM_ID} --cicustom vendor=snippets:snippets/testing.json
```

which will write the ignition json to that file and leave the `user-data` as is FCOS will configure some things via afterburn. Support for vendor-data has been a thing in Proxmox for a few years. https://forum.proxmox.com/threads/combining-custom-cloud-init-with-auto-generated.59008/post-428359

We could check user-data for ignition json first and if it has a cloud-config we can check vendor-data for the ignition data so we don't break things for Flatcar. @abuisine @arcln WDYT?

**Comment by b-** [comment]
> 
> Another thing I noticed in the cidata drive there is a vendor-data file that is empty and you can run 
> 
> ```
> qm set ${VM_ID} --cicustom vendor=snippets:snippets/testing.json
> ```
> 
> which will write the ignition json to that file and leave the `user-data` as is FCOS will configure some things via afterburn. Support for vendor-data has been a thing in Proxmox for a few years. https://forum.proxmox.com/threads/combining-custom-cloud-init-with-auto-generated.59008/post-428359
> 
> We could check user-data for ignition json first and if it has a cloud-config we can check vendor-data for the ignition data so we don't break things for Flatcar. @abuisine @arcln WDYT?

I really like that idea myself!

**Comment by jdoss** [comment]
@b- Something like this https://github.com/coreos/ignition/pull/2014

**Comment by c4rt0** [comment]
This was discussed during today FCOS community meeting.

@jdoss agreed to organize high bandwidth meeting with @jlebon @travier @spresti and @b- to discuss next steps for ignition/afterburn and proxmox static networking

**Comment by jlebon** [comment]
@jdoss @b- and some folks from the CoreOS team met to discuss this. Briefly:
- For the static networking support, we will add another Proxmox support to the OOB network configuration support that already exists in Afterburn (see https://coreos.github.io/afterburn/usage/initrd-network-cmdline/).
- Re. using vendor-data for the Ignition config, we will go ahead with the logic in https://github.com/coreos/ignition/pull/2014. The main concern was whether this conflicts with expectations around vendor-data, or with future plans for Proxmox. We found https://bugzilla.proxmox.com/show_bug.cgi?id=2429 which implies that vendor-data support was added exactly for the purpose of leaving user-data for Proxmox while still allowing true user-specified data to be provided. The expectation is that if in the future Proxmox does surface a UX for user-specified data, it would also use vendor-data.

**Comment by jdoss** [comment]
@dustymabe @travier With https://github.com/coreos/afterburn/pull/1165 approved can we get it merged? Then we need to get releases for Afterburn and Ignition done so they can get included in next `Next` release. 

What other things do we need to do there to get official Proxmox images building?

**Comment by dustymabe** [comment]
I think we are just waiting on the [new afterburn release](https://github.com/coreos/afterburn/issues/1184) to land and then we'd need to add it to our pipeline.

From https://github.com/coreos/fedora-coreos-tracker/issues/1652#issuecomment-2637937326

> We agreed to make proxmox a toplevel supported platform for Fedora CoreOS.

So I'll convert the checklist in the description of this ticket to the more complete version used for a new toplevel platform.

**Comment by spyfly** [comment]
> I think we are just waiting on the [new afterburn release](https://github.com/coreos/afterburn/issues/1184) to land and then we'd need to add it to our pipeline.
New afterburn is released 🎉

**Comment by dustymabe** [comment]
- https://github.com/coreos/fedora-coreos-tracker/issues/1652

**Comment by dustymabe** [comment]
websites work:

https://gitlab.com/fedora/websites-apps/fedora-websites/fedora-websites-3.0/-/merge_requests/1068

**Comment by dustymabe** [comment]
- https://github.com/coreos/coreos-assembler/pull/4142

**Comment by dustymabe** [comment]
- https://github.com/coreos/stream-metadata-go/issues/90

**Comment by dustymabe** [comment]
- https://github.com/coreos/fedora-coreos-stream-generator/issues/73
-

**Comment by jdoss** [comment]
> If anyone is following this and wants to try a dev image.. try out
> 
>     * https://dustymabe.fedorapeople.org/fedora-coreos-42.20250613.dev.1-proxmoxve.x86_64.qcow2.xz
> 
> 
> I don't own a proxmox setup so I have no idea how to deploy this to test it. If you don't know how then you should probably wait until we have docs.

I will test this out later this evening. Thanks Dusty!!

**Comment by b-** [comment]
Me too, this is exciting! 🚀

**Comment by ironicbadger** [comment]
The `.xz` extension doesn't work with the Proxmox CT template URL downloader. But once expanded and imported with `xz -d filename.qcow2.xz` I used the template and the LXC never wanted to create. 

```
Formatting '/var/lib/vz/images/100/vm-100-disk-0.raw', fmt=raw size=8589934592 preallocation=off
Creating filesystem with 2097152 4k blocks and 524288 inodes
Filesystem UUID: e14e4dd9-0238-4468-8e56-573960bb4168
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632
extracting archive '/var/lib/vz/template/cache/fcos.tar.xz'
Total bytes read: 789360640 (753MiB, 181MiB/s)
/etc/os-release file not found and autodetection failed, falling back to 'unmanaged'
implement me in sub-class
 at /usr/share/perl5/PVE/LXC/Setup.pm line 142.
Architecture detection failed: error in setup task (eval)
Falling back to amd64.
Use `pct set VMID --arch ARCH` to change.
```

**Comment by b-** [comment]
> The `.xz` extension doesn't work with the Proxmox CT template URL downloader. But once expanded and imported with `xz -d filename.qcow2.xz` I used the template and the LXC never wanted to create. 
> 
> ```
> Formatting '/var/lib/vz/images/100/vm-100-disk-0.raw', fmt=raw size=8589934592 preallocation=off
> Creating filesystem with 2097152 4k blocks and 524288 inodes
> Filesystem UUID: e14e4dd9-0238-4468-8e56-573960bb4168
> Superblock backups stored on blocks: 
> 	32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632
> extracting archive '/var/lib/vz/template/cache/fcos.tar.xz'
> Total bytes read: 789360640 (753MiB, 181MiB/s)
> /etc/os-release file not found and autodetection failed, falling back to 'unmanaged'
> implement me in sub-class
>  at /usr/share/perl5/PVE/LXC/Setup.pm line 142.
> Architecture detection failed: error in setup task (eval)
> Falling back to amd64.
> Use `pct set VMID --arch ARCH` to change.
> ```

I haven't tried it yet myself, but you're going to want to actually make a VM, _not_ an LXC, and import the disk into that using `qm disk import 123 filename.qcow2 local-zfs --target-disk scsi0`, replacing the filename, storage name, and unused target disk port. Don't forget to `unxz filename.qcow2.xz` first, and you'll also need to put an ignition file as a cloud-init disk as stated further up the issue.

**Comment by dustymabe** [comment]
@jdoss can you PR https://github.com/coreos/fedora-coreos-config/blob/testing-devel/platforms.yaml with correct values for serial console? can you also make sure that if someone doesn't have a console device (i.e. if they didn't start their machine with a console) it still works OK even if the console settings are there on the kernel command line?

**Comment by b-** [comment]
> [@jdoss](https://github.com/jdoss) can you PR [coreos/fedora-coreos-config@`testing-devel`/platforms.yaml](https://github.com/coreos/fedora-coreos-config/blob/testing-devel/platforms.yaml?rgh-link-date=2025-06-20T21%3A38%3A21.000Z) with correct values for serial console? can you also make sure that if someone doesn't have a console device (i.e. if they didn't start their machine with a console) it still works OK even if the console settings are there on the kernel command line?
> 
> Would also be really great if we could link to some docs that mention what the configured console settings should be (if there is such a page).

I'm trying it right now, but I'm 99% certain the reason jdoss's quick script doesn't work OOTB is just because the grub bootloader line doesn't do `console=ttyS0` and service `getty@ttyS0` isn't enabled

**Comment by dustymabe** [comment]
> not doingEquinix Metal mages anymore

Equinix/Packet are part of history now.

**Comment by dustymabe** [comment]
- https://github.com/coreos/fedora-coreos-config/pull/3554
- https://github.com/coreos/fedora-coreos-pipeline/pull/1170

**Comment by kevynb** [comment]
Hello, I’ve been following this from a distance for some time (I would’ve helped test the image but I’m away from laptop until end of June).
This is great work and I thank you all for the effort!

The build pipeline seems to output a .xz image. The proxmox ui and its cli don’t support xz, we have to manually unpack the image before importing it into the proxmox storage.
Would it be feasible to tweak the build pipeline to not compress with xz (sorry, I didn’t research yet what are the supported options, gz is probably supported).
I hesitated from posting this question initially but seeing the fcos browser page that you just updated, I realized that not all released images are compressed using xz (example, the Apple hypervisor one).


This would even enable the automation of deployments through tools like opentofu or terraform. Right now you have to jump through a few hoops: download the image locally, unpack it and then send it to the proxmox server which is not the most optimal flow. You can see an example on the blog post shared above (https://frank.villaro-dixon.eu/posts/coreos-proxmox/). There’s workaround to manage these steps inside open tofu but it still happens on your machine, not the server.

I know that the argument could be made that proxmox should add support for xz compression but given that this pipeline is being tailored specifically for proxmox, I feel this requirement is valid.

I’m ok to work on a PR for that if you feel like you’ve done a lot already (which is true for all the contributors to this issue) but I won’t be able to before mid-July and I’m afraid that once this is closed, we won’t want to change it back.

**Comment by dustymabe** [comment]
Can someone confirm 100% that if the format is `qcow2.gz` it will resolve the problem and make importing much easier?

**Comment by dustymabe** [comment]
If anyone wants to try it and report back their findings:

- https://dustymabe.fedorapeople.org/fedora-coreos-42.20250622.dev.2-proxmoxve.x86_64.qcow2.gz

**Comment by jdoss** [comment]
I'd test it out but I am not exactly sure what @ironicbadger or @b- are talking about since I have always made templates and I download them manually to storage on my cluster. Can we get exact steps of what we need to verify works with compression?

**Comment by kevynb** [comment]
I can be wrong @b- , I based my thinking on the implementation from https://registry.terraform.io/providers/bpg/proxmox/latest/docs/guides/cloud-image which points to https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_download_file that has the compression algorithms and that points to the proxmox api doc on https://pve.proxmox.com/pve-docs/api-viewer/index.html#/nodes/%7Bnode%7D/storage/%7Bstorage%7D/download-url

It’s super hard to read the proxmox api docs on mobile so I might be confusing it with another use case.

The docs from the bpg provider states that the following algorithms are accepted: gz | lzo | zst | bz2 and that ideally the image should be in a .img extension (I base this on the following quote from the docs). The proxmox api doc does not indicate anything.

> Proxmox does not natively support QCOW2 images, but provider can do the conversion for you.

**Comment by kevynb** [comment]
@jdoss @dustymabe sorry I read the error message again and the problem was the output file name I had set, it had no extension. I fixed that (added a `.img` to `coreostest`) and then it worked 👌
So gz is ok!

**Comment by dustymabe** [comment]
@kevynb thanks for the report. A few questions:

1. Does the web UI allow you to import similarly?
2. Once you've imported that image using download-url, can you start a VM and it works?

**Comment by jdoss** [comment]
It also might be worth noting that we can prob use xz as

```
$VZTMPL_EXT_RE_1 = qr/\.tar\.(gz|xz|zst|bz2)/i;
```

has that in the regex too. I will test it later today.

**Comment by kevynb** [comment]
@betermieux I don’t know what you had tried in the past but I had successfully deployed the initial image that @b- had created a few weeks ago using the following terraform config (collated and edited from my phone, so excuse the slight errors you might find, it’s hard 😅)


```hcl

terraform {
  required_providers {
  
  proxmox = {
      source  = "bpg/proxmox"
      version = "0.78.1"
    }
    
    ct = {
      source  = "poseidon/ct"
      version = "0.13.0"
    }
   }
  }


resource "proxmox_virtual_environment_file" "coreos_qcow2" {
  content_type = "iso"
  datastore_id = "my-storage"
  node_name    = "my-node"

# this is because I manually downloaded and renamed the image.
  depends_on = [null_resource.coreos_qcow2]

  source_file {
    path = "/tmp/whatever/fedora-coreos-42.20250512.3.0-proxmoxve.x86_64.qcow2.img"
  }
}


data "ct_config" "gitea_runner_ignition" {
  strict = true
  content = templatefile("butane/gitea-runner-vm.yaml.tftpl", {
    ssh_admin_username   = "username"
    ssh_admin_public_key = file("../ssh_pub_keys/username.pub")
    hostname             = "gitea_runner"
    gitea_runner_registration_token = data.onepassword_item.gitea_runner_registration_token.credential
    gitea_deploy_user = data.onepassword_item.gitea_actions_deploy_credentials.username
    gitea_deploy_pass = data.onepassword_item.gitea_actions_deploy_credentials.credential
  })
}

resource "proxmox_virtual_environment_file" "cloud_user_config" {
  content_type = "snippets"
  datastore_id = "my-storage"
  node_name    = "my-node"

  source_raw {
    data = data.ct_config.gitea_runner_ignition.rendered

    file_name = "gitea_runner.butane-ci-user-data.yml"
  }
}

resource "proxmox_virtual_environment_vm" "gitea_runner_vm" {
  # This must be the name of your Proxmox node within Proxmox
  node_name   = "my-node"
  name        = "gitea-runner"
  description = "Managed by Terraform"

  started = true

  machine = "q35"

  # Since we're installing the guest agent in our Butane config,
  # we should enable it here for better integration with Proxmox
  agent {
    enabled = true
  }

  vga {
    type = "serial0"
  }

  serial_device {
    device = "socket"
  }

  memory {
    dedicated = 2048
  }

  # Here we're referencing the file we uploaded before. Proxmox will
  # clone a new disk from it with the size we're defining.
  disk {
    interface    = "virtio0"
    datastore_id = "my-storage"
    file_id      = proxmox_virtual_environment_file.coreos_qcow2.id
    size         = 32
  }

  # We need a network connection so that we can install the guest agent
  network_device {
    bridge = "vmbr0"
  }

  initialization {
    interface            = "ide2"
    datastore_id         = "my-storage"
    user_data_file_id    = proxmox_virtual_environment_file.cloud_user_config.id
  }
}
```

This got me a working vm that was successfully applying my butane config. Re running the apply was telling me that nothing had changed so everything was working as expected.

As you say, it might not be possible to do in the WebUI yet but the goal of this is automation. I think that if ansible and terraform allow to automate it, it’s already a major step forward.
I’d really like to try deploying it using import instead of iso because the tf provider supports it but I can’t from where I am.

**Comment by dustymabe** [comment]
Let's limit this from rolling into the `stable` stream for now while we workout the image format details here.

The `qcow2.xz` image will be in our `testing`/`next` releases this week, but we can change the format next cycle if there is something better.

**Comment by marmijo** [comment]
The fix for this went into `next` stream release `42.20250623.1.0`. Please try out the new release and report issues.

**Comment by jdoss** [comment]
🔥 🔥 🔥 

Thanks everyone who helped get this going. I am going test asap.

**Comment by dustymabe** [comment]
The big remaining step here is we needs a docs landing page for this.

@jdoss @b- can I lean on you for this?

**Comment by b-** [comment]
I'm not the best at docs, but I wrote a somewhat self-documenting script based on the one @jdoss wrote in [this comment](https://github.com/coreos/fedora-coreos-tracker/issues/1652#issuecomment-2992964247): 

```bash
#!/usr/bin/env bash
set -euo pipefail

### Variables
TEMPLATE_ID=${TEMPLATE_ID:-"10101"}
VM_ID=${VM_ID:-"1002"}
STORAGE=${STORAGE:-"zssd"}
FCOS_STREAM=${FCOS_STREAM:-"testing"}
STREAM_URL=${STREAM_URL:-"https://builds.coreos.fedoraproject.org/streams/${FCOS_STREAM}.json"}
TMPDIR="${TMPDIR:-"/tmp"}"
TMPFILE="${TMPFILE:-$(mktemp "${TMPDIR}/fcos-image-XXXXXX.qcow2")}"
VM_DEFAULTS=(
    --agent 1,fstrim_cloned_disks=1
    --boot order=scsi0
    --ide2 "${STORAGE}:cloudinit"
    --memory 2048
    --name fedora-coreos-${FCOS_STREAM}-template
    --net0 virtio,bridge=vmbr0
    --scsi0 "${STORAGE}:0,import-from=${TMPFILE}"
    --scsihw virtio-scsi-single
    --serial0 socket
    --vga serial0
)

# Flag to ensure cleanup only runs once
CLEANED_UP=0

cleanup() {
  if (( CLEANED_UP )); then
    return
  fi
  CLEANED_UP=1

  echo "Performing cleanup..."
  rm -f "${TMPFILE}"
}

# Trap normal and abnormal exits
trap cleanup EXIT INT TERM HUP ERR

### Functions
function check_requirements() {
    requirements=(
        "curl"
        "jq"
        "xzcat"
    )
    function _exists() {
        local cmd="$1"
        command -v "${cmd}" > /dev/null 2>&1
    }
    function _msg_missing() {
        local cmd="$1"
        echo "Error: ${cmd} is not installed"
        echo "Required commands: ${requirements[*]}"
    }
    for cmd in "${requirements[@]}"; do
        # shellcheck disable=SC2310 # invoked in an 'if' condition
        if ! _exists "${cmd}"; then
            _msg_missing "${cmd}"
            exit 1
        fi
    done
}
check_requirements

fetch() {
  local url="$1"
  local out="${2:-/dev/stdout}"

  # Default curl flags
  local curl_opts=(
    --fail
    --show-error
    --connect-timeout 5
    --max-time 60
    --location
  )
  # Try to get Content-Length
  local size
  size=$(curl -sfI -L "$url" | awk -F': ' 'tolower($1) == "content-length" { print $2; exit }' | tr -d '\r')

  # Default curl flags
  local curl_opts=(
    --fail
    --location
    --silent
    --show-error
    --connect-timeout 5
    --max-time 60
  )

  if command -v pv >/dev/null && [[ "$size" =~ ^[0-9]+$ ]]; then
    echo "Downloading with pv (size: $size bytes)" >&2
    curl "${curl_opts[@]}" "$url" | pv -s "$size" > "$out"
  else
    echo "Downloading with curl progress meter" >&2
    curl "${curl_opts[@]}" "$url" -# -o "$out"
  fi

}

function fetch_metadata_item() {
    # we use := to assign the value of fetch("$url") to the var
    local url="${STREAM_URL}"
    json="${JSON:=$(fetch "$url" | jq -cr )}"
    local architecture="${ARCHITECTURE:-"x86_64"}"
    local platform="${PLATFORM:-"proxmoxve"}"
    local format="${FORMAT:-"qcow2.xz"}"
    local query_base='.architectures[$arch].artifacts[$plat]'
    local item="$1" # release
    local query="${query_base}.${item}"
    export json
    echo "${json}" | jq -r --arg arch "$architecture" --arg plat "$platform" --arg format "$format" "$query"
}

function download_fcos() {
    RELEASE="${RELEASE:=$(fetch_metadata_item "release")}"
    LOCATION="${LOCATION:=$(fetch_metadata_item 'formats[$format].disk.location')}"
    IMAGE_XZ="$(basename "${LOCATION}")"
    fetch "${LOCATION}" | xzcat > "${TMPFILE}"
}

function stop_vm() {
    local vmid="$1"
    if [[ "$2" == "force" ]]; then
        qm stop "${vmid}" || true
    else
        qm stop "${vmid}"
    fi
}

function delete_vm() {
    local vmid="$1"
    if [[ "$2" == "force" ]]; then
        stop_vm "${vmid}" force
        qm destroy "${vmid}" || true
    else
        qm destroy "${vmid}"
    fi
}

function create_vm() {
    local settings=("${VM_DEFAULTS[@]}")
    qm create "${TEMPLATE_ID}" "${settings[@]}" --name fedora-coreos-${FCOS_STREAM}-${RELEASE}-template
}

### Main
delete_vm "${VM_ID}" force
delete_vm "${TEMPLATE_ID}" force

download_fcos

create_vm

qm template "${TEMPLATE_ID}"

qm clone "${TEMPLATE_ID}" "${VM_ID}" --name fcos-stable-test-"${VM_ID}"
qm set "${VM_ID}" --sshkey "${HOME}/.ssh/authorized_keys" --cicustom vendor=snippets:snippets/testing.json --ciupgrade 0
#qm set "${VM_ID}" --sshkey "${HOME}/.ssh/authorized_keys" --ciupgrade 0
qm resize "${VM_ID}" scsi0 +90G

qm start "${VM_ID}"
qm terminal "${VM_ID}"
```

Everything works at first glance, including serial terminal. Maybe someone else can draft some docs and I can look them over and test?

**Comment by jdoss** [comment]
I will draft some docs over the weekend. @dustymabe what repo do I PR them against again?

**Comment by dustymabe** [comment]
https://github.com/coreos/fedora-coreos-docs

**Comment by dustymabe** [comment]
Hey All, It's been a few weeks. I'd like to promote this so we have `stable` proxmoxve images. A couple things:

1. Did we decide if changing to `.gz` would help anything?
2. Would be really nice if we had docs when it goes to `stable`. @jdoss will you be able to work on this soon?

**Comment by b-** [comment]
I'm not convinced that switching to .gz will help anything -- whether users provision using the API/Terraform or using shell scripts or what have you, the xz image works just fine.

I spent some time learning enough Terraform to hurt myself, so I wrote a simple/hacky/dirty TF module that deploys CoreOS on Proxmox using this new platform. It needs a lot of work before it's something I'd really share _publicly_ (as in, outside of this GitHub issue or online chat rooms), but I thought I should link it here anyway: https://github.com/perchnet/p/tree/main/modules/coreos-vm

**Comment by g5pw** [comment]
@dustymabe now that I think about it, I might have started it without a SSH key the first time, but I've subsequently added it and powered the machine off/on. SSH keys were ultimately applied, as I was able to ssh in with the key.

**Comment by marshallford** [comment]
The stable build (42.20250705.3.0) released a couple of days ago is working and can be downloaded compressed by Proxmox using Terraform (as others have mentioned). No need to resize the disk either! Thank you @dustymabe!

```
resource "proxmox_virtual_environment_download_file" "fcos" {
  content_type            = "iso"
  datastore_id            = var.proxmox_datastore_directory
  file_name               = "fedora-coreos-42.20250705.3.0-proxmoxve.x86_64.img"
  node_name               = var.proxmox_node
  url                     = "https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/42.20250705.3.0/x86_64/fedora-coreos-42.20250705.3.0-proxmoxve.x86_64.qcow2.xz"
  checksum                = "dab2cfafe397aa96e2885d11ab89a1463bc0dd49a04c7f06bfef7246d13f0437"
  checksum_algorithm      = "sha256"
  decompression_algorithm = "zst"
}
```

Full example: https://github.com/marshallford/terraform-provider-ansible/blob/main/examples/complete/proxmox/main.tf

**Comment by dustymabe** [comment]
As @marshallford mentioned:

The fix for this went into `stable` stream release `42.20250705.3.0`.

**Comment by dustymabe** [comment]
Docs PR https://github.com/coreos/fedora-coreos-docs/pull/742 merged!
